MR取得周りを見直す

- OpenなMRだけを取得する
- 日付フィールドでエラー
```java
Caused by: org.springframework.web.client.RestClientException: Error while extracting response for type [class [Lio.github.tttol.mrls.dto.GitLabMergeRequestApiResponseDto;] and content type [application/json]
        at org.springframework.web.client.HttpMessageConverterExtractor.extractData(HttpMessageConverterExtractor.java:118) ~[spring-web-6.0.3.jar:6.0.3]
        at org.springframework.web.client.RestTemplate$ResponseEntityResponseExtractor.extractData(RestTemplate.java:1132) ~[spring-web-6.0.3.jar:6.0.3]
        at org.springframework.web.client.RestTemplate$ResponseEntityResponseExtractor.extractData(RestTemplate.java:1115) ~[spring-web-6.0.3.jar:6.0.3]
        at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:865) ~[spring-web-6.0.3.jar:6.0.3]
        at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:704) ~[spring-web-6.0.3.jar:6.0.3]
        at io.github.tttol.mrls.external.GitLabApiExecutor.getMergeRequests(GitLabApiExecutor.java:41) ~[classes/:na]
        ... 53 common frames omitted
Caused by: org.springframework.http.converter.HttpMessageNotReadableException: JSON parse error: Cannot deserialize value of type `java.time.LocalDateTime` from String "2023-01-30T11:25:28.681+09:00": Failed to deserialize java.time.LocalDateTime: (java.time.format.DateTimeParseException) Text '2023-01-30T11:25:28.681+09:00' could not be parsed, unparsed text found at index 23
        at org.springframework.http.converter.json.AbstractJackson2HttpMessageConverter.readJavaType(AbstractJackson2HttpMessageConverter.java:406) ~[spring-web-6.0.3.jar:6.0.3]
        at org.springframework.http.converter.json.AbstractJackson2HttpMessageConverter.read(AbstractJackson2HttpMessageConverter.java:354) ~[spring-web-6.0.3.jar:6.0.3]
        at org.springframework.web.client.HttpMessageConverterExtractor.extractData(HttpMessageConverterExtractor.java:103) ~[spring-web-6.0.3.jar:6.0.3]
        ... 58 common frames omitted
Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.time.LocalDateTime` from String "2023-01-30T11:25:28.681+09:00": Failed to deserialize java.time.LocalDateTime: (java.time.format.DateTimeParseException) Text '2023-01-30T11:25:28.681+09:00' could not be parsed, unparsed text found at index 23
 at [Source: (org.springframework.util.StreamUtils$NonClosingInputStream); line: 1, column: 165] (through reference chain: java.lang.Object[][0]->io.github.tttol.mrls.dto.GitLabMergeRequestApiResponseDto["created_at"])
        at com.fasterxml.jackson.databind.exc.InvalidFormatException.from(InvalidFormatException.java:67) ~[jackson-databind-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.databind.DeserializationContext.weirdStringException(DeserializationContext.java:1996) ~[jackson-databind-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.databind.DeserializationContext.handleWeirdStringValue(DeserializationContext.java:1224) ~[jackson-databind-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.datatype.jsr310.deser.JSR310DeserializerBase._handleDateTimeException(JSR310DeserializerBase.java:176) ~[jackson-datatype-jsr310-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.datatype.jsr310.deser.LocalDateTimeDeserializer._fromString(LocalDateTimeDeserializer.java:179) ~[jackson-datatype-jsr310-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.datatype.jsr310.deser.LocalDateTimeDeserializer.deserialize(LocalDateTimeDeserializer.java:81) ~[jackson-datatype-jsr310-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.datatype.jsr310.deser.LocalDateTimeDeserializer.deserialize(LocalDateTimeDeserializer.java:40) ~[jackson-datatype-jsr310-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129) ~[jackson-databind-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:392) ~[jackson-databind-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:185) ~[jackson-databind-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.databind.deser.std.ObjectArrayDeserializer.deserialize(ObjectArrayDeserializer.java:216) ~[jackson-databind-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.databind.deser.std.ObjectArrayDeserializer.deserialize(ObjectArrayDeserializer.java:26) ~[jackson-databind-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323) ~[jackson-databind-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:2105) ~[jackson-databind-2.14.1.jar:2.14.1]
        at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1481) ~[jackson-databind-2.14.1.jar:2.14.1]
        at org.springframework.http.converter.json.AbstractJackson2HttpMessageConverter.readJavaType(AbstractJackson2HttpMessageConverter.java:395) ~[spring-web-6.0.3.jar:6.0.3]
        ... 60 common frames omitted
Caused by: java.time.format.DateTimeParseException: Text '2023-01-30T11:25:28.681+09:00' could not be parsed, unparsed text found at index 23
        at java.base/java.time.format.DateTimeFormatter.parseResolved0(Unknown Source) ~[na:na]
        at java.base/java.time.format.DateTimeFormatter.parse(Unknown Source) ~[na:na]
        at java.base/java.time.LocalDateTime.parse(Unknown Source) ~[na:na]
        at com.fasterxml.jackson.datatype.jsr310.deser.LocalDateTimeDeserializer._fromString(LocalDateTimeDeserializer.java:177) ~[jackson-datatype-jsr310-2.14.1.jar:2.14.1]
        ... 71 common frames omitted
```

--- a/src/main/java/io/github/tttol/mrls/dto/GitLabMergeRequestApiResponseDto.java
+++ b/src/main/java/io/github/tttol/mrls/dto/GitLabMergeRequestApiResponseDto.java
@@ -8,7 +8,7 @@ import lombok.Data;
 import lombok.NoArgsConstructor;
 import org.springframework.format.annotation.DateTimeFormat;
 
-import java.time.LocalDateTime;
+import java.time.OffsetDateTime;
 import java.util.List;
 
 @Data
@@ -30,22 +30,22 @@ public class GitLabMergeRequestApiResponseDto {
     String state;
 
     @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
-    LocalDateTime createdAt;
+    OffsetDateTime createdAt;
 
     @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
-    LocalDateTime updatedAt;
+    OffsetDateTime updatedAt;
 
     String mergedBy;
 
     String mergeUser;
 
     @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
-    LocalDateTime mergedAt;
+    OffsetDateTime mergedAt;
 
     String closedBy;
 
     @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
-    LocalDateTime closedAt;
+    OffsetDateTime closedAt;
 
     String targetBranch;
 

--- a/src/main/java/io/github/tttol/mrls/form/MrDetailForm.java
+++ b/src/main/java/io/github/tttol/mrls/form/MrDetailForm.java
@@ -1,7 +1,7 @@
 package io.github.tttol.mrls.form;
 
-import java.time.LocalDateTime;
+import java.time.OffsetDateTime;
 
-public record MrDetailForm(String title, String webUrl, UserForm author, LocalDateTime createdAt,
-                           LocalDateTime updatedAt) {
+public record MrDetailForm(String title, String webUrl, UserForm author, OffsetDateTime createdAt,
+                           OffsetDateTime updatedAt) {
 }

--- a/src/main/java/io/github/tttol/mrls/service/MergeRequestService.java
+++ b/src/main/java/io/github/tttol/mrls/service/MergeRequestService.java
@@ -5,75 +5,69 @@ import io.github.tttol.mrls.external.GitLabApiExecutor;
 import io.github.tttol.mrls.form.MrDetailForm;
 import io.github.tttol.mrls.form.MrInfoForm;
 import io.github.tttol.mrls.form.UserForm;
-import java.time.LocalDateTime;
-import java.time.ZoneId;
-import java.time.ZonedDateTime;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.time.ZoneOffset;
 import java.util.Comparator;
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.stream.Collectors;
-import lombok.RequiredArgsConstructor;
-import org.springframework.stereotype.Service;
 
 @Service
 @RequiredArgsConstructor
 public class MergeRequestService {
 
-  private final GitLabApiExecutor gitLabApiExecutor;
+    private final GitLabApiExecutor gitLabApiExecutor;
 
-  public List<MrInfoForm> get(String accessToken) {
-    var mergeRequestInfoDtos = executeGitLabApi(accessToken);
-    return mergeRequestInfoDtos.stream()
-        .collect(Collectors.groupingBy(
-                e -> Objects.isNull(e.getAssignee()) ?
-                    e.getAuthor().getId() : e.getAssignee().getId()
-            )
-        )
-        .values().stream().map(this::generateForm)
-        .sorted(Comparator.comparing(e -> e.assignee().id()))
-        .toList();
-  }
-
-  private List<GitLabMergeRequestApiResponseDto> executeGitLabApi(String accessToken) {
-    return gitLabApiExecutor.getMergeRequests(accessToken);
-  }
-
-  private MrInfoForm generateForm(List<GitLabMergeRequestApiResponseDto> responseDtos) {
-    var responseDto = responseDtos.stream().findAny().orElseThrow();
-    var assigneeForm = Optional.ofNullable(responseDto.getAssignee())
-        .map(assignee -> new UserForm(
-                assignee.getId(),
-                assignee.getUsername(),
-                assignee.getName(),
-                assignee.getState(),
-                assignee.getAvatarUrl(),
-                assignee.getWebUrl()
-            )
-        ).orElse(UserForm.empty());
-    var linkForms = responseDtos.stream().map(e ->
-        new MrDetailForm(
-            e.getTitle(),
-            e.getWebUrl(),
-            Optional.ofNullable(e.getAuthor())
-                .map(assignee -> new UserForm(
-                        assignee.getId(),
-                        assignee.getUsername(),
-                        assignee.getName(),
-                        assignee.getState(),
-                        assignee.getAvatarUrl(),
-                        assignee.getWebUrl()
-                    )
+    public List<MrInfoForm> get(final String accessToken) {
+        final var mergeRequestInfoDtos = executeGitLabApi(accessToken);
+        return mergeRequestInfoDtos.stream()
+                .collect(Collectors.groupingBy(
+                                e -> Objects.isNull(e.getAssignee()) ?
+                                        e.getAuthor().getId() : e.getAssignee().getId()
+                        )
                 )
-                .orElse(UserForm.empty()),
-            toJst(e.getCreatedAt()),
-            toJst(e.getUpdatedAt())
-        )).toList();
-    return new MrInfoForm(assigneeForm, linkForms, linkForms.size());
-  }
+                .values().stream().map(this::generateForm)
+                .sorted(Comparator.comparing(e -> e.assignee().id()))
+                .toList();
+    }
 
-  private LocalDateTime toJst(LocalDateTime utc) {
-    return ZonedDateTime.of(utc, ZoneId.of("UTC"))
-        .withZoneSameInstant(ZoneId.of("Asia/Tokyo")).toLocalDateTime();
-  }
+    private List<GitLabMergeRequestApiResponseDto> executeGitLabApi(final String accessToken) {
+        return gitLabApiExecutor.getMergeRequests(accessToken);
+    }
+
+    private MrInfoForm generateForm(final List<GitLabMergeRequestApiResponseDto> responseDtos) {
+        final var responseDto = responseDtos.stream().findAny().orElseThrow();
+        final var assigneeForm = Optional.ofNullable(responseDto.getAssignee())
+                .map(assignee -> new UserForm(
+                                assignee.getId(),
+                                assignee.getUsername(),
+                                assignee.getName(),
+                                assignee.getState(),
+                                assignee.getAvatarUrl(),
+                                assignee.getWebUrl()
+                        )
+                ).orElse(UserForm.empty());
+        final var linkForms = responseDtos.stream().map(e ->
+                new MrDetailForm(
+                        e.getTitle(),
+                        e.getWebUrl(),
+                        Optional.ofNullable(e.getAuthor())
+                                .map(assignee -> new UserForm(
+                                                assignee.getId(),
+                                                assignee.getUsername(),
+                                                assignee.getName(),
+                                                assignee.getState(),
+                                                assignee.getAvatarUrl(),
+                                                assignee.getWebUrl()
+                                        )
+                                )
+                                .orElse(UserForm.empty()),
+                        e.getCreatedAt().withOffsetSameInstant(ZoneOffset.of("+09:00")),
+                        e.getUpdatedAt().withOffsetSameInstant(ZoneOffset.of("+09:00"))
+                )).toList();
+        return new MrInfoForm(assigneeForm, linkForms, linkForms.size());
+    }
 }

--- a/src/main/resources/application.yml
+++ b/src/main/resources/application.yml
@@ -5,7 +5,7 @@ app:
       id: ${GITLAB_PROJECT_ID}
       accessToken: ${GITLAB_ACCESS_TOKEN}
     api:
-      endpoint: https://${app.gitlab.host}/api/v4/projects/${app.gitlab.project.id}/merge_requests
+      endpoint: https://${app.gitlab.host}/api/v4/projects/${app.gitlab.project.id}/merge_requests?state=opened
 logging:
   level:
     root: INFO

--- a/src/main/resources/templates/fragment/footer.html
+++ b/src/main/resources/templates/fragment/footer.html
@@ -3,7 +3,7 @@
     <body>
         <footer th:fragment="footer">
             <div class="footerBox">
-                <span class="appVer">version  1.0.0</span>
+                <span class="appVer">version  1.1.3</span>
             </div>
         </footer>
     </body>

--- a/src/test/java/io/github/tttol/mrls/controller/MrListControllerTest.java
+++ b/src/test/java/io/github/tttol/mrls/controller/MrListControllerTest.java
@@ -12,7 +12,8 @@ import org.springframework.test.web.servlet.MockMvc;
 import org.springframework.test.web.servlet.setup.MockMvcBuilders;
 import org.springframework.web.context.WebApplicationContext;
 
-import java.time.LocalDateTime;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
 import java.util.List;
 
 import static org.mockito.ArgumentMatchers.anyString;
@@ -36,13 +37,13 @@ public class MrListControllerTest {
     public void init() {
         mockMvc = MockMvcBuilders.webAppContextSetup(this.webApplicationContext).build();
 
-        var assignee = UserDto.builder()
+        final var assignee = UserDto.builder()
                 .id(1)
                 .username("assignee_username1")
                 .name("assignee_name1")
                 .state("active")
                 .build();
-        var mergeRequestInfoDtos = List.of(
+        final var mergeRequestInfoDtos = List.of(
                 GitLabMergeRequestApiResponseDto.builder()
                         .assignee(assignee)
                         .author(UserDto.builder()
@@ -53,8 +54,8 @@ public class MrListControllerTest {
                                 .build())
                         .webUrl("url1")
                         .title("title1")
-                        .createdAt(LocalDateTime.of(1970, 1, 1, 0, 0, 0))
-                        .updatedAt(LocalDateTime.of(2000, 1, 1, 0, 0, 0))
+                        .createdAt(OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC))
+                        .updatedAt(OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC))
                         .build(),
                 GitLabMergeRequestApiResponseDto.builder()
                         .assignee(assignee)
@@ -66,8 +67,8 @@ public class MrListControllerTest {
                                 .build())
                         .webUrl("url2")
                         .title("title2")
-                        .createdAt(LocalDateTime.of(1970, 1, 1, 0, 0, 0))
-                        .updatedAt(LocalDateTime.of(2000, 1, 1, 0, 0, 0))
+                        .createdAt(OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC))
+                        .updatedAt(OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC))
                         .build(),
                 GitLabMergeRequestApiResponseDto.builder()
                         .assignee(assignee)
@@ -79,8 +80,8 @@ public class MrListControllerTest {
                                 .build())
                         .webUrl("url3")
                         .title("title3")
-                        .createdAt(LocalDateTime.of(1970, 1, 1, 0, 0, 0))
-                        .updatedAt(LocalDateTime.of(2000, 1, 1, 0, 0, 0))
+                        .createdAt(OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC))
+                        .updatedAt(OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, ZoneOffset.UTC))
                         .build()
         );
         doReturn(mergeRequestInfoDtos).when(gitLabApiExecutor).getMergeRequests(anyString());

--- a/src/test/java/io/github/tttol/mrls/service/MergeRequestServiceTest.java
+++ b/src/test/java/io/github/tttol/mrls/service/MergeRequestServiceTest.java
@@ -1,201 +1,199 @@
 package io.github.tttol.mrls.service;
 
-import static org.assertj.core.api.Assertions.assertThat;
-import static org.mockito.ArgumentMatchers.anyString;
-import static org.mockito.Mockito.doReturn;
-
 import io.github.tttol.mrls.dto.GitLabMergeRequestApiResponseDto;
 import io.github.tttol.mrls.dto.UserDto;
 import io.github.tttol.mrls.external.GitLabApiExecutor;
 import io.github.tttol.mrls.form.MrDetailForm;
 import io.github.tttol.mrls.form.MrInfoForm;
 import io.github.tttol.mrls.form.UserForm;
-import java.time.LocalDateTime;
-import java.util.List;
-import org.junit.jupiter.api.AfterEach;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.DisplayName;
-import org.junit.jupiter.api.Nested;
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.*;
 import org.mockito.InjectMocks;
 import org.mockito.Mock;
 import org.mockito.MockitoAnnotations;
 
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.doReturn;
+
 public class MergeRequestServiceTest {
 
-  private AutoCloseable closeable;
+    private AutoCloseable closeable;
 
-  @BeforeEach
-  void setUp() {
-    closeable = MockitoAnnotations.openMocks(this);
-  }
-
-  @AfterEach
-  void close() throws Exception {
-    closeable.close();
-  }
-
-  @InjectMocks
-  private MergeRequestService mergeRequestService;
-
-  @Mock
-  private GitLabApiExecutor gitLabApiExecutor;
-
-  @Nested
-  class Get {
-
-    @Test
-    @DisplayName("GitLabからMRを取得しdto->formの変換ができること")
-    void formTest() {
-      var assignee = UserDto.builder()
-          .id(1)
-          .username("assignee_username1")
-          .name("assignee_name1")
-          .state("active")
-          .build();
-      var mergeRequestInfoDtos = List.of(
-          GitLabMergeRequestApiResponseDto.builder()
-              .assignee(assignee)
-              .author(UserDto.builder()
-                  .id(11)
-                  .username("author_username11")
-                  .name("author_name11")
-                  .state("active")
-                  .build())
-              .webUrl("url1")
-              .title("title1")
-              .createdAt(LocalDateTime.of(1970, 1, 1, 0, 0, 0))
-              .updatedAt(LocalDateTime.of(2000, 1, 1, 0, 0, 0))
-              .build(),
-          GitLabMergeRequestApiResponseDto.builder()
-              .assignee(assignee)
-              .author(UserDto.builder()
-                  .id(11)
-                  .username("author_username12")
-                  .name("author_name12")
-                  .state("active")
-                  .build())
-              .webUrl("url2")
-              .title("title2")
-              .createdAt(LocalDateTime.of(1970, 1, 1, 0, 0, 0))
-              .updatedAt(LocalDateTime.of(2000, 1, 1, 0, 0, 0))
-              .build(),
-          GitLabMergeRequestApiResponseDto.builder()
-              .assignee(assignee)
-              .author(UserDto.builder()
-                  .id(13)
-                  .username("author_username13")
-                  .name("author_name13")
-                  .state("active")
-                  .build())
-              .webUrl("url3")
-              .title("title3")
-              .createdAt(LocalDateTime.of(1970, 1, 1, 0, 0, 0))
-              .updatedAt(LocalDateTime.of(2000, 1, 1, 0, 0, 0))
-              .build()
-      );
-      doReturn(mergeRequestInfoDtos).when(gitLabApiExecutor).getMergeRequests(anyString());
-      var expected = List.of(new MrInfoForm(
-          new UserForm(1, "assignee_username1", "assignee_name1", "active", null, null),
-          List.of(
-              new MrDetailForm(
-                  "title1",
-                  "url1",
-                  new UserForm(
-                      11,
-                      "author_username11",
-                      "author_name11",
-                      "active",
-                      null,
-                      null
-                  ),
-                  LocalDateTime.of(1970, 1, 1, 9, 0, 0),
-                  LocalDateTime.of(2000, 1, 1, 9, 0, 0)
-              ),
-              new MrDetailForm(
-                  "title2",
-                  "url2",
-                  new UserForm(
-                      11,
-                      "author_username12",
-                      "author_name12",
-                      "active",
-                      null,
-                      null
-                  ),
-                  LocalDateTime.of(1970, 1, 1, 9, 0, 0),
-                  LocalDateTime.of(2000, 1, 1, 9, 0, 0)
-              ),
-              new MrDetailForm(
-                  "title3",
-                  "url3",
-                  new UserForm(
-                      13,
-                      "author_username13",
-                      "author_name13",
-                      "active",
-                      null,
-                      null
-                  ),
-                  LocalDateTime.of(1970, 1, 1, 9, 0, 0),
-                  LocalDateTime.of(2000, 1, 1, 9, 0, 0)
-
-              )
-          ),
-          3
-      ));
-
-      var actual = mergeRequestService.get(anyString());
-
-      assertThat(actual).usingRecursiveComparison().isEqualTo(expected);
+    @BeforeEach
+    void setUp() {
+        closeable = MockitoAnnotations.openMocks(this);
     }
 
-    @Test
-    @DisplayName("MRのAssigneeが未指定の場合")
-    void noAssigneeTest() {
-      var mergeRequestInfoDtos = List.of(
-          GitLabMergeRequestApiResponseDto.builder()
-              .assignee(null)
-              .author(UserDto.builder()
-                  .id(11)
-                  .username("author_username11")
-                  .name("author_name11")
-                  .state("active")
-                  .build())
-              .webUrl("url1")
-              .title("title1")
-              .createdAt(LocalDateTime.of(1970, 1, 1, 0, 0, 0))
-              .updatedAt(LocalDateTime.of(2000, 1, 1, 0, 0, 0))
-              .build()
-      );
-      doReturn(mergeRequestInfoDtos).when(gitLabApiExecutor).getMergeRequests(anyString());
-      var expected = List.of(new MrInfoForm(
-          UserForm.empty(),
-          List.of(
-              new MrDetailForm(
-                  "title1",
-                  "url1",
-                  new UserForm(11, "author_username11", "author_name11", "active", null, null),
-                  LocalDateTime.of(1970, 1, 1, 9, 0, 0),
-                  LocalDateTime.of(2000, 1, 1, 9, 0, 0)
-              )
-          ),
-          1
-      ));
-
-      var actual = mergeRequestService.get(anyString());
-
-      assertThat(actual).usingRecursiveComparison().isEqualTo(expected);
+    @AfterEach
+    void close() throws Exception {
+        closeable.close();
     }
 
-    @Test
-    @DisplayName("MR0件の場合")
-    void noMrTest() {
-      doReturn(List.of()).when(gitLabApiExecutor).getMergeRequests(anyString());
-      var expected = List.of();
-      var actual = mergeRequestService.get(anyString());
+    @InjectMocks
+    private MergeRequestService mergeRequestService;
 
-      assertThat(actual).usingRecursiveComparison().isEqualTo(expected);
+    @Mock
+    private GitLabApiExecutor gitLabApiExecutor;
+
+    @Nested
+    class Get {
+
+        @Test
+        @DisplayName("GitLabからMRを取得しdto->formの変換ができること")
+        void formTest() {
+            final var assignee = UserDto.builder()
+                    .id(1)
+                    .username("assignee_username1")
+                    .name("assignee_name1")
+                    .state("active")
+                    .build();
+            final var mergeRequestInfoDtos = List.of(
+                    GitLabMergeRequestApiResponseDto.builder()
+                            .assignee(assignee)
+                            .author(UserDto.builder()
+                                    .id(11)
+                                    .username("author_username11")
+                                    .name("author_name11")
+                                    .state("active")
+                                    .build())
+                            .webUrl("url1")
+                            .title("title1")
+                            .createdAt(OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, ZoneOffset.of("+09:00")))
+                            .updatedAt(OffsetDateTime.of(2000, 1, 1, 0, 0, 0, 0, ZoneOffset.of("+09:00")))
+                            .build(),
+                    GitLabMergeRequestApiResponseDto.builder()
+                            .assignee(assignee)
+                            .author(UserDto.builder()
+                                    .id(11)
+                                    .username("author_username12")
+                                    .name("author_name12")
+                                    .state("active")
+                                    .build())
+                            .webUrl("url2")
+                            .title("title2")
+                            .createdAt(OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, ZoneOffset.of("+09:00")))
+                            .updatedAt(OffsetDateTime.of(2000, 1, 1, 0, 0, 0, 0, ZoneOffset.of("+09:00")))
+                            .build(),
+                    GitLabMergeRequestApiResponseDto.builder()
+                            .assignee(assignee)
+                            .author(UserDto.builder()
+                                    .id(13)
+                                    .username("author_username13")
+                                    .name("author_name13")
+                                    .state("active")
+                                    .build())
+                            .webUrl("url3")
+                            .title("title3")
+                            .createdAt(OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, ZoneOffset.of("+09:00")))
+                            .updatedAt(OffsetDateTime.of(2000, 1, 1, 0, 0, 0, 0, ZoneOffset.of("+09:00")))
+                            .build()
+            );
+            doReturn(mergeRequestInfoDtos).when(gitLabApiExecutor).getMergeRequests(anyString());
+            final var expected = List.of(new MrInfoForm(
+                    new UserForm(1, "assignee_username1", "assignee_name1", "active", null, null),
+                    List.of(
+                            new MrDetailForm(
+                                    "title1",
+                                    "url1",
+                                    new UserForm(
+                                            11,
+                                            "author_username11",
+                                            "author_name11",
+                                            "active",
+                                            null,
+                                            null
+                                    ),
+                                    OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, ZoneOffset.of("+09:00")),
+                                    OffsetDateTime.of(2000, 1, 1, 0, 0, 0, 0, ZoneOffset.of("+09:00"))
+                            ),
+                            new MrDetailForm(
+                                    "title2",
+                                    "url2",
+                                    new UserForm(
+                                            11,
+                                            "author_username12",
+                                            "author_name12",
+                                            "active",
+                                            null,
+                                            null
+                                    ),
+                                    OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, ZoneOffset.of("+09:00")),
+                                    OffsetDateTime.of(2000, 1, 1, 0, 0, 0, 0, ZoneOffset.of("+09:00"))
+                            ),
+                            new MrDetailForm(
+                                    "title3",
+                                    "url3",
+                                    new UserForm(
+                                            13,
+                                            "author_username13",
+                                            "author_name13",
+                                            "active",
+                                            null,
+                                            null
+                                    ),
+                                    OffsetDateTime.of(1970, 1, 1, 0, 0, 0, 0, ZoneOffset.of("+09:00")),
+                                    OffsetDateTime.of(2000, 1, 1, 0, 0, 0, 0, ZoneOffset.of("+09:00"))
+
+                            )
+                    ),
+                    3
+            ));
+
+            final var actual = mergeRequestService.get(anyString());
+
+            assertThat(actual).usingRecursiveComparison().isEqualTo(expected);
+        }
+
+        @Test
+        @DisplayName("MRのAssigneeが未指定の場合")
+        void noAssigneeTest() {
+            final var mergeRequestInfoDtos = List.of(
+                    GitLabMergeRequestApiResponseDto.builder()
+                            .assignee(null)
+                            .author(UserDto.builder()
+                                    .id(11)
+                                    .username("author_username11")
+                                    .name("author_name11")
+                                    .state("active")
+                                    .build())
+                            .webUrl("url1")
+                            .title("title1")
+                            .createdAt(OffsetDateTime.of(1970, 1, 1, 9, 0, 0, 0, ZoneOffset.of("+09:00")))
+                            .updatedAt(OffsetDateTime.of(2000, 1, 1, 9, 0, 0, 0, ZoneOffset.of("+09:00")))
+                            .build()
+            );
+            doReturn(mergeRequestInfoDtos).when(gitLabApiExecutor).getMergeRequests(anyString());
+            final var expected = List.of(new MrInfoForm(
+                    UserForm.empty(),
+                    List.of(
+                            new MrDetailForm(
+                                    "title1",
+                                    "url1",
+                                    new UserForm(11, "author_username11", "author_name11", "active", null, null),
+                                    OffsetDateTime.of(1970, 1, 1, 9, 0, 0, 0, ZoneOffset.of("+09:00")),
+                                    OffsetDateTime.of(2000, 1, 1, 9, 0, 0, 0, ZoneOffset.of("+09:00"))
+                            )
+                    ),
+                    1
+            ));
+
+            final var actual = mergeRequestService.get(anyString());
+
+            assertThat(actual).usingRecursiveComparison().isEqualTo(expected);
+        }
+
+        @Test
+        @DisplayName("MR0件の場合")
+        void noMrTest() {
+            doReturn(List.of()).when(gitLabApiExecutor).getMergeRequests(anyString());
+            final var expected = List.of();
+            final var actual = mergeRequestService.get(anyString());
+
+            assertThat(actual).usingRecursiveComparison().isEqualTo(expected);
+        }
     }
-  }
 }

